### PR TITLE
Merge tables plants and plant_detail and make its data available in frontend.

### PR DESCRIPTION
**Pull request by badnames**
This pull request makes the scraped data by @aidnurs available in the "Add seeds" dialog.
For this purpose, plant_detail now replaces the old plants table.
A new variety / species search was added by @Bushuo.

## Basics

- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildserver is happy.
- [x] The PR is rebased with current master.

## Checklist

- [x] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation
- [x] I fixed all affected decisions
- [x] I added code comments, logging, and assertions as appropriate
- [x] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is conforming to [our Documentation Guidelines](/doc/documentation.md)
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to our Coding Guidelines
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

### CODE REVIEW

**Comment by markus2330** [comment]
Great to see this PR! Seems like it doesn't build at the moment.

> This pull request makes the scraped data by @aidnurs available in the "Add plants" dialog.

Do you mean "Add seeds" here? Adding new plants is not part of this milestone.

File: backend/src/model/entity/plants_impl.rs [hunk 1]
```diff
@@ -14,16 +16,43 @@ impl Plants {
...
+        let query_with_placeholders = format!("%{capitalized_query}%");
+
+        let query = plants::table
+            .select(all_columns)
+            .filter(
+                sql::<Bool>("LOWER(binomial_name) LIKE ")
+                    .bind::<Text, _>(&query_with_placeholders)
+                    .sql(" OR LOWER(ARRAY_TO_STRING(common_name, ' ')) LIKE ")
+                    .bind::<Text, _>(&query_with_placeholders),
+            )
```
Comment (Benjaminimal) [inline backend/src/model/entity/plants_impl.rs:43]: If we have a powerful ORM we should use its features.
Only if we can't express our statement or they would be inefficient should we fall back to raw SQL.

There is no ned for `to_lowercase`, use ilike instead of like.
For filtering on the binomial name you can do this `.filter(binomail_name.ilike(query_with_placeholders))`.
https://docs.diesel.rs/master/diesel/expression_methods/trait.PgTextExpressionMethods.html#method.ilike

I'm not sure how to achieve something similar for the values in `common_name` but it can probably be done with the querybuilder as well.

File: backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql [hunk 2]
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE seeds DROP COLUMN plant_id;
+
+DROP TABLE plants;
+
+ALTER TABLE plant_detail RENAME TO plants;
+ALTER TABLE seeds ADD COLUMN plant_id INTEGER REFERENCES plants(id);
+
+INSERT INTO plants (id, binomial_name, common_name) VALUES (-1, 'Testia testia', ARRAY['Testplant']);
```
Comment (Benjaminimal) [inline backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql:8]: Test data shouldn't be part of migrations.
Please set up the test data in the test where you need it.
You can use this function to make sure nothing will be committed during a test to ensure test isolation.
https://docs.diesel.rs/master/diesel/connection/trait.Connection.html#method.test_transaction

A quick search led me to this example of how to use it.
https://github.com/diesel-rs/diesel/issues/1549#issuecomment-621792203

**Review by Benjaminimal — COMMENTED** [review]
Great Job!
This should give us a clean start for all the further use cases.

File: backend/src/model/entity/plants_impl.rs (hunk 1 above)
Comment (badnames) [reply backend/src/model/entity/plants_impl.rs:43]: I opted to use raw SQL strings because I could not get the "correct" diesel ORM-methods to work.
Tbh. I can't quite remember what the exact problem was, but there was a trait missing somewhere. 

That being sain, I agree with your points and would prefer these changes if someone could get them to work.

File: backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql (hunk 2 above)
Comment (badnames) [reply backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql:8]: How are you finding these things when I've been sitting here googling this stuff for hours :sweat_smile:....

Anyways, thanks for the suggestion :+1: .

Comment (Benjaminimal) [reply backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql:8]: :shrug: 
Here is another example that is closer to our set up.
https://cloudmaker.dev/actix-integration-tests/

But maybe we can leave it for now and create a new issue to set up backend tests in a usable way?

File: backend/src/model/entity/plants_impl.rs (hunk 1 above)
Comment (Bushuo) [reply backend/src/model/entity/plants_impl.rs]: Did it :)

File: frontend/src/features/seeds/routes/CreateSeed.tsx [hunk 3]
```diff
@@ -2,20 +2,48 @@ import PageLayout from '../../../components/Layout/PageLayout';
...
+
+    const timeout = setTimeout(() => {
+      searchFunction(searchParam);
+    }, delay);
+
+    setTimeoutId(timeout);
+  };
+
+  return [debouncedSearch];
+}
```
Comment (buenaflor) [inline frontend/src/features/seeds/routes/CreateSeed.tsx:28]: Debouncing might be useful for other components, maybe put this in the global `utils` folder

**Review by buenaflor — CHANGES_REQUESTED** [review]

File: frontend/src/features/seeds/routes/CreateSeed.tsx (hunk 3 above)
Comment (Bushuo) [reply frontend/src/features/seeds/routes/CreateSeed.tsx]: I agree.
I also changed the hook, so that it can be easily integrated into react-query, once we merge this.

File: backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql (hunk 2 above)
Comment (markus2330) [reply backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql:8]: Did you already see the tests we already have in backend/src/test/seed.rs?

Btw. @Benjaminimal **excellent** review.

Comment (markus2330) [reply backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql:8]: I agree there should be further tests, within M1.

File: backend/src/controller/plants.rs [hunk 4]
```diff
@@ -1,25 +1,49 @@
...
-pub async fn find_all(pool: Data<Pool>) -> Result<HttpResponse> {
+pub async fn find_all_or_search(
+    query: Option<web::Query<PlantsSearchParameters>>,
+    pool: Data<Pool>,
+) -> Result<HttpResponse> {
+    if let Some(query) = query {
+        search(query, pool).await
+    } else {
+        find_all(pool).await
+    }
```
Comment (Benjaminimal) [inline backend/src/controller/plants.rs:32]: It might be cleaner if we don't call controllers from another controller.
You can directly go to the service layer like this.
```suggestion
    let response = match query {
        Some(parameters) => {
            web::block(move || service::plants::search(&pool, &parameters)).await??
        }
        None => web::block(move || service::plants::find_all(&pool)).await??,
    };
    Ok(HttpResponse::Ok().json(response))
```
You can also get rid of the other controllers.

File: backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql (hunk 2 above)
Comment (Benjaminimal) [reply backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql:8]: The seeds tests do the same.
Adding test data in migrations.
https://github.com/ElektraInitiative/PermaplanT/blob/9ac1b16809e94aa9cc6c5d8fadf41f8f74d4c249/backend/migrations/2023-01-21-024920_create_seeds/up.sql#L19-L22

No test transactions in the seeds test.
https://github.com/ElektraInitiative/PermaplanT/blob/9ac1b16809e94aa9cc6c5d8fadf41f8f74d4c249/backend/src/test/seed.rs

We also have a lot of code duplication here at the start of each test function.
This should be pulled out into some setup which also sets up a test database (maybe with some fixtures) so we can use test transactions easily.
We would also need a tear-down to drop this database after the last test has run.
I looked into it briefly and it seems like rust's built-in test crate doesn't provide easy setup and tear-down out of the box.
@kitzbergerg 
Do you know any recipe or good test crate to achieve this?

Comment (Bushuo) [reply backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql]: I added a test utility that initializes the pool so that it uses the test transactions.
This should be enough to cover easy test cases.
I made a comment about its limitations which @badnames and I identified.

Comment (kitzbergerg) [reply backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql]: As far as I know the tests in Rust shouldn't actually be used how we use them. They should be more on the side of unit tests rather than integration tests.
I just found a [crate](https://docs.rs/mockall/latest/mockall/) to mock traits (~interfaces). I think we should think about tests more carefully soon (especially as we currently require a running db that is not part of the backend) and split into unit and integration tests.

**Comment by kitzbergerg** [comment]
That's a bit more general, but in my opinion this PR is way to large. With 51 changed files there is way to much to think about and it is quite difficult to read/see if there are any errors anywhere.

I haven't looked at the code in detail though, maybe there is a good reason/it is only possible in a large PR like this. 

Generally I would however prefer if PR wouldn't get that large. Maybe split the tasks into smaller subtasks next time. That way it is easier to write and review and you don't have huge merge conflicts with other branches before/after merging.

File: backend/src/model/dto.rs [hunk 5]
```diff
@@ -53,11 +53,17 @@ pub struct NewSeedDto {
 
 #[allow(clippy::missing_docs_in_private_items)] // TODO: See #97.
```
Comment (kitzbergerg) [inline backend/src/model/dto.rs:54]: Please remove lines `#[allow(clippy::missing_docs_in_private_items)] // TODO: See #97.` completely for all newly created enums/structs.
Instead you should document the structs and fields with code comments e.g. 
```rust
/// This represents ... and is used for ...
#[typeshare]
#[derive(Debug, Serialize, PartialEq, Eq, Deserialize, ToSchema)]
pub struct PlantsSummaryDto {
    /// The id of the entity
    pub id: i32,
   /// <Explaination of the field>
    pub binomial_name: String,
   /// <Explaination of the field>
    pub common_name: Option<Vec<Option<String>>>,
}
```

Please write comments that can be used to fully understand what the structs/fields do and what might be special about them. Those comments will later on be part of the documentation (generated from code).

You can look at the [doc of Vec](https://doc.rust-lang.org/std/vec/struct.Vec.html) and the [corresponding code](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#151) as and example. The doc is generated from code.

**Comment by badnames** [comment]
> That's a bit more general, but in my opinion this PR is way to large. With 51 changed files there is way to much to think about and it is quite difficult to read/see if there are any errors anywhere.
> 
> I haven't looked at the code in detail though, maybe there is a good reason/it is only possible in a large PR like this.
> 
> Generally I would however prefer it if PRs don't get that large. Maybe split the tasks into smaller subtasks next time. That way it is easier to write and review and you don't have huge merge conflicts with other branches before/after merging.

Yes I agree.

The problem was that the goal posts moved quite a bit over time. At first we only wanted to add pagination for the existing plants-typeahead (hence the branch name). Then we discovered that the old plants table is no longer needed.

Since this is my first time working with actix-web and diesel I greatly underestimated the amount of work needed to merge these two tables.
While adding tests, we discovered that some refactoring was needed in out unit testing system (thanks again to @Benjaminimal).
And, since master was quite far ahead of our branch by this point, a lot of files needed to be changed to fix all merge conflicts.

So to summarize, let us all agree to only address one issue per pull request from now on.
I will therefore exclude the changes to our schema requested by @Benjaminimal and add them as a separate issue (again many thanks for pointing this out).

File: backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql (hunk 2 above)
Comment (markus2330) [reply backend/migrations/2023-03-29-111855_merge_plant_tables/up.sql]: Yes, excellent suggestion, we definitely should separate unit and integration tests. There is #39, can you please kick this off as part of #158?

**Comment by markus2330** [comment]
jenkins build please

**Review by buenaflor — APPROVED** [review]
👍

**Review by markus2330 — APPROVED** [review]
Amazing job! Great to finally see a search for plants. :1st_place_medal:
